cvat-ui/src/components/annotation-page/annotation-filter/annotation-filter-item/annotation-filter-item.tsx

```diff
@@ -0,0 +1,115 @@
+// Copyright (C) 2020-2021 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+
+import { Tag } from 'antd';
+import PropTypes from 'prop-types';
+import React, { ReactElement, useEffect, useReducer } from 'react';
+
+interface State {
+    id: string;
+    concatenator: string;
+    filterBy: string;
+    operator: string;
+    value: string;
+    attribute: string;
+    attributeOperator: string;
+    attributeValue: string;
+    anotherAttributeLabel: string;
+    anotherAttributeValue: string;
+    left: string[];
+    right: string[];
+}
+
+interface Props {
+    item: State;
+    onEdit: Function;
+    onDelete: Function;
+    onGrouping: Function;
+}
+
+enum ActionType {
+    addLeft,
+    addRight,
+    removeLeft,
+    removeRight,
+}
+
+const reducer = (state: State, action: { type: ActionType; payload?: any }): State => {
+    switch (action.type) {
+        case ActionType.addLeft:
+            return { ...state, left: [...state.left, '('] };
+        case ActionType.addRight:
+            return { ...state, right: [...state.right, ')'] };
+        case ActionType.removeLeft:
+            state.left.pop();
```

Comment:
state.right property before reducing is equal to state.right property after.
It is better to use something destructive like that:
``return { ...state, left: state.left.slice(0, -1) };``